immer/detail/type_traits.hpp

```diff
@@ -0,0 +1,158 @@
+    std::true_type {};
+
+template <typename T>
+constexpr bool is_dereferenceable = is_dereferenceable_t<T>::value;
+
+template <typename T, typename = void>
+struct provides_preincrement_t : std::false_type {};
+
+template <typename T>
+struct provides_preincrement_t
+<T, std::enable_if_t
+ <std::is_same<T&, decltype(++(std::declval<T&>()))>::value>> :
+    std::true_type {};
+
+template <typename T>
+constexpr bool provides_preincrement = provides_preincrement_t<T>::value;
+
+using std::swap;
+
+template <typename T, typename U, typename = void>
+struct is_swappable_with_t : std::false_type {};
+
+// Does not account for non-referenceable types
+template <typename T, typename U>
+struct is_swappable_with_t
+<T, U, void_t<decltype(swap(std::declval<T&>(), std::declval<U&>())),
+              decltype(swap(std::declval<U&>(), std::declval<T&>()))>> :
+    std::true_type {};
+
+template <typename T>
+constexpr bool is_swappable = is_swappable_with_t<T&, T&>::value;
+
+template <typename T, typename = void>
+struct is_iterator_t : std::false_type {};
+
+// See http://en.cppreference.com/w/cpp/concept/Iterator
+template <typename T>
+struct is_iterator_t
+<T, void_t
+ <std::enable_if_t
+  <provides_preincrement<T>
+   and is_dereferenceable<T>
+   // accounts for non-referenceable types
+   and std::is_copy_constructible<T>::value
+   and std::is_copy_assignable<T>::value
+   and std::is_destructible<T>::value
+   and is_swappable<T>>,
+  typename std::iterator_traits<T>::value_type,
+  typename std::iterator_traits<T>::difference_type,
+  typename std::iterator_traits<T>::reference,
+  typename std::iterator_traits<T>::pointer,
+  typename std::iterator_traits<T>::iterator_category>> :
+    std::true_type {};
+
+template<typename T>
+constexpr bool is_iterator = is_iterator_t<T>::value;
+
+template<typename T, typename U=T, typename = void>
+struct equality_comparable_t : std::false_type {};
+
+template<typename T, typename U>
+struct equality_comparable_t
+<T, U, std::enable_if_t
+ <std::is_same
+  <bool, decltype(std::declval<T&>() == std::declval<U&>())>::value>> :
+    std::true_type {};
+
+template<typename T, typename U=T>
+constexpr bool equality_comparable = equality_comparable_t<T, U>::value;
+
+template<typename T, typename U=T, typename = void>
+struct inequality_comparable_t : std::false_type {};
+
+template<typename T, typename U>
+struct inequality_comparable_t
+<T, U, std::enable_if_t
+ <std::is_same
+  <bool, decltype(std::declval<T&>() == std::declval<U&>())>::value>> :
+    std::true_type {};
+
+template<typename T, typename U=T>
+constexpr bool inequality_comparable =
+  inequality_comparable_t<T, U>::value;
+
+template<typename T, typename U, typename = void>
+struct compatible_sentinel_t : std::false_type {};
+
+template<typename T, typename U>
+struct compatible_sentinel_t
+<T, U, std::enable_if_t
+ <is_iterator<T>
+  and equality_comparable<T, U>
+  and inequality_comparable<T, U>>> :
```

Comment:
I have mixed feelings about requiring inequality checks for sentinels. What do other people do? In particular, does range v3 require `!=` for sentinels?

Reply:
Whether it is strictly required, I'll have to double-check with one of the range-v3 library authors. I can say with certainty that a defined Iterator/Sentinel inequality operation is assumed in at least some subset of the range-v3 algorithm implementation. See the following [example](https://github.com/ericniebler/range-v3/blob/8ccd974b5cbb91dc9de7ab969abd568fcf569019/include/range/v3/utility/iterator.hpp#L274).

Reply:
That makes sense, thanks for the link!